### PR TITLE
refact(CORE/Mail): Avoid big mail loading and saving time by only loading 50 mails at a time.

### PR DESCRIPTION
**Pull request by peti446**
## CHANGES PROPOSED:
-  Changed the way mail is handled to only have 50 mail loaded from the DB when the client request mail update to keep load time low and avoid crashes with big amount of mails.
-  Save changed mail when new mail is fetched as now the mail size is much lower and can be done without problems

## TESTS PERFORMED:
- Build on windows with no errors
- Created a character with 150K mails to test loading time when mail is opened
- Send mail between characters, from console and AH with no apparent errors

## HOW TO TEST THE CHANGES:

- Use the mail system as much as possible with different characters online and offline. Including AH.
- Send many mails to a character to check load time

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -955,7 +954,7 @@ Player::~Player()
         delete itr->second;
 
     //all mailed items should be deleted, also all mail should be deallocated
-    for (PlayerMails::iterator itr = m_mail.begin(); itr != m_mail.end(); ++itr)
+    for (PlayerMails::iterator itr = m_mailCache.begin(); itr != m_mailCache.end(); ++itr)
         delete *itr;
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:959]:
```suggestion
    for (PlayerMails::iterator itr = m_mailCache.begin(); itr != m_mailCache.end(); ++itr)
    {
        delete *itr;
    }

```

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -3634,26 +3633,39 @@ void Player::SendNewMail()
...
+        PreparedQueryResult result = CharacterDatabase.Query(stmt);
+        if (result)
         {
-            if (!m_nextMailDelivereTime || m_nextMailDelivereTime > (*itr)->deliver_time)
-                m_nextMailDelivereTime = (*itr)->deliver_time;
+            Field* fields = result->Fetch();
+            m_nextMailDelivereTime = fields[0].GetUInt64();
+        }
+    }
+    {
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:3650]: is this curly brackets necessary?

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -4567,7 +4579,7 @@ void Player::SetFreeTalentPoints(uint32 points)
 
 Mail* Player::GetMail(uint32 id)
 {
-    for (PlayerMails::iterator itr = m_mail.begin(); itr != m_mail.end(); ++itr)
+    for (PlayerMails::iterator itr = m_mailCache.begin(); itr != m_mailCache.end(); ++itr)
         if ((*itr)->messageID == id)
             return (*itr);
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:4585]:
```suggestion
    for (PlayerMails::iterator itr = m_mailCache.begin(); itr != m_mailCache.end(); ++itr)
    {
        if ((*itr)->messageID == id)
        {
            return (*itr);
        }
    }
```

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -18957,108 +18973,33 @@ void Player::_LoadMailInit(PreparedQueryResult resultUnread, PreparedQueryResult
...
-
-                AddMItem(item);
-            }
-
-            prevMailID = fields[14].GetUInt32();
-        }
-        while (result->NextRow());
+    //In case we are not expecting to receive new mail we just exits, not really necessary to refresh any data
+    if (m_nextMailDelivereTime > time(nullptr))
+        return;
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:18981]:
```suggestion
    if (m_nextMailDelivereTime > time(nullptr))
    {
        return;
    }

```

File: src/server/game/Entities/Player/Player.cpp [hunk 5]
```diff
@@ -19996,12 +19936,13 @@ void Player::_SaveInventory(SQLTransaction& trans)
 
 void Player::_SaveMail(SQLTransaction& trans)
 {
-    if (!m_mailsLoaded)
+    if (!GetMailCacheSize() || !m_mailsUpdated)
         return;
+
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:19942]:
```suggestion
    if (!GetMailCacheSize() || !m_mailsUpdated)
    {
        return;
    }    

```

File: src/server/game/Mails/Mail.cpp [hunk 6]
```diff
@@ -245,43 +245,35 @@ void MailDraft::SendMailTo(SQLTransaction& trans, MailReceiver const& receiver,
...
-                    pReceiver->AddMItem(mailItemIter->second);
-            }
-        }
-        else if (!m_items.empty())
+        if (!m_items.empty())
         {
-            SQLTransaction temp = SQLTransaction(nullptr);
-            deleteIncludedItems(temp);
+            for (MailItemMap::iterator mailItemIter = m_items.begin(); mailItemIter != m_items.end(); ++mailItemIter)
+                pReceiver->AddMItem(mailItemIter->second);
```
Comment (Helias) [inline src/server/game/Mails/Mail.cpp:276]:
```suggestion
            for (MailItemMap::iterator mailItemIter = m_items.begin(); mailItemIter != m_items.end(); ++mailItemIter)
            {
                pReceiver->AddMItem(mailItemIter->second);
            }
```

**Review by Helias — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (peti446) [reply src/server/game/Entities/Player/Player.cpp:3659]: No not really, changed it :)

**Review by Helias — APPROVED** [review]

**Comment by Helias** [comment]
Need to fix the build

**Comment by peti446** [comment]
Should be done with next commit no idea how the line got duplicated 🤔

**Comment by FrancescoBorzi** [comment]
@peti446 there are conflicts

**Comment by locus313** [comment]
@peti446 I am working on testing this. Could you provide the details of how you created the 150k mail on a character? (not tried to do that before)

**Comment by peti446** [comment]
@locus313 I used a soap script running in the background for a while while I was playing around:
`
<?php

$conn = new SoapClient(NULL, array(
    'location' => 'http://127.0.0.1:7878/',
    'uri' => 'urn:AC',
    'style' => SOAP_RPC,
    'login' => 'test2',
    'password' => 'test'
));

function generateRandomString($length = 10) {
    $characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
    $charactersLength = strlen($characters);
    $randomString = '';
    for ($i = 0; $i < $length; $i++) {
        $randomString .= $characters[rand(0, $charactersLength - 1)];
    }
    return $randomString;
}
            
for($x = 0; $x <= 120000; $x++)
{
    echo $conn->executeCommand(new SoapParam('.send mail Tseae "'.generateRandomString(10).'" "'. generateRandomString(150) .'"', 'command'));
}
 `

**Comment by locus313** [comment]
@peti446 I used your script above without any problems. But I did find that when I login to the character that was the target for the generated mails. The worldserver crashes after a few seconds. I am using docker on centos 7 for my testing.

**Comment by peti446** [comment]
@locus313 Thanks will have a look on Linux see if I can reproduce the crash. Did you see any error message before the crash ?

**Comment by locus313** [comment]
@peti446 did not see any error message before the crash.

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers can anyone please check this?

**Comment by peti446** [comment]
The crash @locus313 referred should now be fixed.

**Comment by locus313** [comment]
Just tested. No crash this time :). Look good.

**Pull request merged**